edgePy/DGEList.py

```diff
@@ -272,37 +306,120 @@ def library_size(self) -> np.array:
+
+    def cpm(self, transform_to_log: bool = False, prior_count: float = PRIOR_COUNT) -> "DGEList":
+        """Normalize the DGEList to read counts per million."""
+        counts = 1e6 * self.counts / np.sum(self.counts, axis=0)
+        current_log = self.log
+        if transform_to_log:
+            counts = self.log_transform(counts, prior_count)
+            current_log = True
+            
+        return self.copy(counts=counts, current_log=current_log)    
 
     def rpkm(
-        self, gene_lengths: Mapping, log: bool = False, prior_count: float = PRIOR_COUNT
-    ) -> None:
+        self, gene_data: ImportCanonicalData, log: bool = False, prior_count: float = PRIOR_COUNT
+    ) -> "DGEList":
         """Return the DGEList normalized to reads per kilobase of gene length
-        per million reads.
+        per million reads. (RPKM =   numReads / ( geneLength/1000 * totalNumReads/1,000,000 )
 
         """
-        raise NotImplementedError
-
-        # TODO: Implement here
-        # self = self.cpm(log=log, prior_count=prior_count)
+        current_log = self.log
+        temp_gene_len = []
+
+        if self.log:
+            self.counts = np.exp(self.counts)
+            current_log = False
+
+        gene_mask = []
+        gene_ensg = []
+        col_sum = np.sum(self.counts, axis=0)
+
+        for gene in self.genes:
+            if gene.startswith("ENSG"):
+                gene_name = gene
+                gene_ensg.append(gene_name)
+                if gene_data.has_gene(gene_name):
+                    gene_mask.append(True)
+                    temp_gene_len.append(
+                        gene_data.get_length_of_canonical_transcript(gene_name) / 1e3
+                    )
+                else:
+                    gene_mask.append(False)
+            else:
+                t_gene = gene_data.get_genes_from_symbol(gene)
+                if t_gene:
+                    if len(t_gene) > 1:
+                        gene_name = gene_data.pick_gene_id(t_gene)
+                    else:
+                        gene_name = t_gene[0]
+                    gene_ensg.append(gene_name)
+                    if gene_data.has_gene(gene_name):
+                        gene_mask.append(True)
+                        temp_gene_len.append(
+                            gene_data.get_length_of_canonical_transcript(gene_name) / 1e3
+                        )
+                    else:
+                        gene_mask.append(False)
+                else:
+                    gene_mask.append(False)
+
+        genes = self.genes[gene_mask].copy()
+        counts = self.counts[gene_mask].copy()
+
+        counts = (counts.T / temp_gene_len).T
+        counts = counts / (col_sum / 1e6)
+
+        current_log = self.log
+        if transform_to_log:
+            counts = self.log_transform(counts, prior_count)
+            current_log = True
+
+        return self.copy(counts=counts, current_log=current_log, genes=genes)
```

Comment:
Setting genes here is counter intuitive in my opinion (i.e. a user does not know, that the genes table is being updated)